x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/state/CategorizerStats.java

```diff
@@ -0,0 +1,359 @@
+            ignoreUnknownFields, a -> new Builder((String) a[0]));
+
+        parser.declareString(ConstructingObjectParser.constructorArg(), Job.ID);
+        parser.declareString((modelSizeStat, s) -> {}, Result.RESULT_TYPE);
+        parser.declareString(Builder::setPartitionFieldName, PARTITION_FIELD_NAME);
+        parser.declareString(Builder::setPartitionFieldValue, PARTITION_FIELD_VALUE);
+        parser.declareLong(Builder::setCategorizedDocCount, CATEGORIZED_DOC_COUNT_FIELD);
+        parser.declareLong(Builder::setTotalCategoryCount, TOTAL_CATEGORY_COUNT_FIELD);
+        parser.declareLong(Builder::setFrequentCategoryCount, FREQUENT_CATEGORY_COUNT_FIELD);
+        parser.declareLong(Builder::setRareCategoryCount, RARE_CATEGORY_COUNT_FIELD);
+        parser.declareLong(Builder::setDeadCategoryCount, DEAD_CATEGORY_COUNT_FIELD);
+        parser.declareLong(Builder::setFailedCategoryCount, FAILED_CATEGORY_COUNT_FIELD);
+        parser.declareField(Builder::setCategorizationStatus,
+            p -> CategorizationStatus.fromString(p.text()), CATEGORIZATION_STATUS_FIELD, ValueType.STRING);
+        parser.declareField(Builder::setLogTime,
+            p -> TimeUtils.parseTimeFieldToInstant(p, LOG_TIME_FIELD.getPreferredName()), LOG_TIME_FIELD, ValueType.VALUE);
+        parser.declareField(Builder::setTimestamp,
+            p -> TimeUtils.parseTimeFieldToInstant(p, TIMESTAMP_FIELD.getPreferredName()), TIMESTAMP_FIELD, ValueType.VALUE);
+
+        return parser;
+    }
+
+    private final String jobId;
+    private final String partitionFieldName;
+    private final String partitionFieldValue;
+    private final long categorizedDocCount;
+    private final long totalCategoryCount;
+    private final long frequentCategoryCount;
+    private final long rareCategoryCount;
+    private final long deadCategoryCount;
+    private final long failedCategoryCount;
+    private final CategorizationStatus categorizationStatus;
+    private final Instant timestamp;
+    private final Instant logTime;
+
+    private CategorizerStats(String jobId, @Nullable String partitionFieldName, @Nullable String partitionFieldValue,
+                             long categorizedDocCount, long totalCategoryCount, long frequentCategoryCount, long rareCategoryCount,
+                             long deadCategoryCount, long failedCategoryCount, CategorizationStatus categorizationStatus, Instant timestamp,
+                             Instant logTime) {
+        this.jobId = Objects.requireNonNull(jobId);
+        this.partitionFieldName = partitionFieldName;
+        this.partitionFieldValue = partitionFieldValue;
+        this.categorizedDocCount = categorizedDocCount;
+        this.totalCategoryCount = totalCategoryCount;
+        this.frequentCategoryCount = frequentCategoryCount;
+        this.rareCategoryCount = rareCategoryCount;
+        this.deadCategoryCount = deadCategoryCount;
+        this.failedCategoryCount = failedCategoryCount;
```

Comment:
We don't check that all of these are `> 0`. This is probably OK since we are getting this data from C++ and there is an implicit trust.

Reply:
Yes, good point.  This is dodgy when we're using `VLong` serialization.  It means someone can DoS us by updating a document to contain a negative number.

This same problem applies to most of our other results classes too, so I think it's best to deal with it in a separate PR.

Like you say, the C++ won't send negative values for unsigned counters, so it's not a likely problem.